### PR TITLE
Add AWS SES connector with API v2

### PR DESCRIPTION
**Pull request by kasthuriraajan**
Connector supports AWS SES API v2.

## Automation tests
 - Unit tests 
   > Unit test is added for each remote functions.
 - Integration tests
   > N/A

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes
 - Ran FindSecurityBugs plugin and verified report? no
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes

OS - Ubuntu 20.04
Ballerina Swan Lake Beta 3

https://docs.aws.amazon.com/ses/latest/APIReference-V2/Welcome.html
https://docs.aws.amazon.com/general/latest/gr/signing_aws_api_requests.html

### CODE REVIEW

File: ses/Ballerina.toml [hunk 1]
```diff
@@ -0,0 +1,11 @@
+[package]
+org = "ballerinax"
+name = "aws.ses"
+version = "1.0.0-SNAPSHOT"
+license= ["Apache-2.0"]
+authors = ["Ballerina"]
+keywords = ["aws.ses", "amazon", "aws", "ses", "email"]
```
Comment (LakshanSS) [inline ses/Ballerina.toml:7]: Need to add keywords in the latest format

File: ses/Package.md [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+* Perform AWS SES related operations programmatically. The `ballerinax/aws.ses` module provides this capability.
+
+### Compatibility
+|                    | Version         |  
+|--------------------|-----------------|
+| Ballerina Language | Swan Lake Beta3 |
+| Amazon SES API     | v2              |
+
+## Report issues
+To report bugs, request new features, start new discussions, view project boards, etc., go to the [Ballerina AWS SES repository](https://github.com/ballerina-platform/module-ballerinax-aws.ses)
```
Comment (LakshanSS) [inline ses/Package.md:15]: Let's add the following text here

To report bugs, request new features, start new discussions, view project boards, etc., go to the [Ballerina Extended Library repository](https://github.com/ballerina-platform/ballerina-extended-library)

File: ses/client.bal [hunk 3]
```diff
@@ -0,0 +1,519 @@
...
+    # + return - `http:Error` in case of failure to initialize or `null` if successfully initialized
+    public isolated function init(ClientConfiguration amazonSesconfiguration, http:ClientConfiguration httpConfig = {}) 
+                                  returns error? {
+        self.accessKey = amazonSesconfiguration.awsCredentials.accessKeyId;
+        self.secretKey = amazonSesconfiguration.awsCredentials.secretAccessKey;
+        self.securityToken = amazonSesconfiguration.awsCredentials?.securityToken;
+        self.region = amazonSesconfiguration.region;
+        self.host = AWS_SERVICE + DOT + self.region + DOT+ AWS_HOST;
+        string endpoint = HTTPS + self.host;
+        self.awsSesClient = check new (endpoint,httpConfig);
```
Comment (LakshanSS) [inline ses/client.bal:53]: Need a space after `endpoint,`
self.awsSesClient = check new (endpoint, httpConfig);

File: ses/records.bal [hunk 4]
```diff
@@ -0,0 +1,539 @@
...
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+# Represents AWS client configuration.
+#
+# + awsCredentials - AWS credentials  
+# + region - AWS region  
+public type ClientConfiguration record {
```
Comment (LakshanSS) [inline ses/records.bal:21]: Need to change the name as `ConnectionConfig`

In all the connectors, we are using this same name

File: ses/utils.bal [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+import ballerina/crypto;
+import ballerina/jballerina.java;
+import ballerina/http;
+import ballerina/url;
+import ballerina/regex;
+
+isolated function getSignedRequestHeaders(string host, string accessKey, string secretKey, string region, string verb, 
+                             string uri, string payload) returns map<string>|error {
+
+            string content_type =APPLICATION_JSON;
```
Comment (LakshanSS) [inline ses/utils.bal:28]: Check formatting

**Comment by LakshanSS** [comment]
Dependencies.toml is missing

**Comment by LakshanSS** [comment]
Need to pack the svg icon within the resources folder

File: ses/client.bal [hunk 6]
```diff
@@ -0,0 +1,519 @@
...
+    }
+
+    # Creates a contact list.
+    #
+    # + contactListCreationRequest - The request payload to create contact list
+    # + return - An empty json on success else an error
+    @display {label: "Create Contact List"}
+    remote isolated function createContactList(@display {label: "Contact List To Create"}
+                                               ContactListCreationRequest contactListCreationRequest)
+                                               returns @display {label: "Contact List Creation Response"} json|error {
```
Comment (Chuhaa) [inline ses/client.bal:63]: Can't we return as a record instead of json?

File: ses/client.bal [hunk 7]
```diff
@@ -0,0 +1,519 @@
...
+    # + return - An aws.ses:Contact record on success else an error
+    @display {label: "Get Contact"}
+    remote isolated function getContact(@display {label: "Contact List Name"} string contactListName,
+                                        @display {label: "Email Address"} string emailAddress)
+                                        returns @display {label: "Contact"} Contact| error {
+        string requestUri = URI + "contact-lists/" + contactListName + "/contacts/"+ emailAddress;
+        map<string> signedRequestHeaders = check getSignedRequestHeaders(self.host, self.accessKey, self.secretKey,
+                                                                         self.region, GET, requestUri, EMPTY_STRING);
+        http:Response httpResponse = check self.awsSesClient->get(requestUri, signedRequestHeaders);
+        return (check handleHttpResponse(httpResponse)).cloneWithType(Contact);
```
Comment (Chuhaa) [inline ses/client.bal:185]:
```suggestion
        return self.awsSesClient->get(requestUri, signedRequestHeaders, targetType = Contact);
```
Can't we do in this way?

File: ses/constants.bal [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+const AWS_HOST = "amazonaws.com";
+const AWS_SERVICE = "email";
+const string DEFAULT_REGION = "us-east-1";
+const string URI = "/v2/email/";
```
Comment (Chuhaa) [inline ses/constants.bal:20]:
```suggestion
const string AWS_HOST = "amazonaws.com";
const string AWS_SERVICE = "email";
const string DEFAULT_REGION = "us-east-1";
const string URI = "/v2/email/";
```

More readable

File: ses/tests/test.bal [hunk 9]
```diff
@@ -0,0 +1,452 @@
...
+}
+function testGetContactList() {
+    log:printInfo("Testing GetContactList");
+    ContactList|error response =  amazonSesClient-> getContactList(contactListName);
+    if response is ContactList {
+        test:assertEquals(response?.ContactListName, contactListName, "Expected contact-list is not obtained");
+    } else {
+        test:assertFail(response.toString());
+    }
+}
```
Comment (Chuhaa) [inline ses/tests/test.bal:90]:
```suggestion
function testGetContactList() returns error? {
    log:printInfo("Testing GetContactList");
    ContactList|error response =  amazonSesClient-> getContactList(contactListName);
    test:assertEquals(response?.ContactListName, contactListName, "Expected contact-list is not obtained");
}
```

File: ses/client.bal [hunk 10]
```diff
@@ -0,0 +1,519 @@
...
+    #
+    # + contactListName - The name of the contact list
+    # + contactCreationRequest - The request payload to create contact
+    # + return - An empty json on success else an error
+    @display {label: "Create Contact"}
+    remote isolated function createContact(@display {label: "Contact List Name"} string contactListName,
+                                           @display {label: "Request To Create Contact"}
+                                           ContactCreationRequest contactCreationRequest) 
+                                           returns @display {label: "Contact Creation Response"} json|error {
+        string requestUri = URI + "contact-lists/" + contactListName + "/contacts";
```
Comment (Chuhaa) [inline ses/client.bal:141]: Shall we create a const variable for `"contact-lists/"` and `"/contacts"` and reuse them? It is used in many places

File: ses/Ballerina.toml (hunk 1 above)
Comment (LakshanSS) [reply ses/Ballerina.toml:7]:
```suggestion
keywords = ["communication/email", "cost/freemium", "vendor/amazon"]
```

File: ses/client.bal (hunk 6 above)
Comment (kasthuriraajan) [reply ses/client.bal]: It is removed since it's an empty response. Error is only returned.

File: ses/client.bal (hunk 7 above)
Comment (kasthuriraajan) [reply ses/client.bal]: Resolved in an alternative way which is discussed in code review.

**Pull request merged**